### PR TITLE
Use same 'build-logic' for 'hedera-node' and 'platform-sdk'

### PR DESCRIPTION
**Pull request by jjohannes**
**Description**:

This is the first step to integrate the 'hedera-node' and 'platform-sdk' builds.
Work was done together with @hendrikebbers.

-  'platform-sdk' now also uses the Gradle convention plugins from `build-logic` – most duplication has been removed
- While doing the integration, we found some things we needed to improve – see other comments

Next steps after this PR:
1. Remove remaining duplications by looking in detail at remaining `com.swirlds.platform.*.gradle.kts` convention plugins
2. Also use `com.hedera.hashgraph.jpms-module-dependencies` dependency setup, "Gradle platform project" in  'platform-sdk' and get rid of duplicated version definitions (currently these are still in the version catalog in `platform-sdk/settings.gradle.kts`)

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -35,6 +35,9 @@ plugins {
 
 includeBuild("hedera-platform")
 
+// If you turn this is on, 'platform-sdk' modules are built from source
+// includeBuild("platform-sdk")
```
Comment (jjohannes) [inline settings.gradle.kts:39]: Adding this line will _include_ `platform-sdk` into the `hedera-node` build. Then all modules needed will be built from source automatically when you work on `hedera-node`. The `swirldsVersion` in `hedera-platform/build.gradle.kts` is ignored then.

File: platform-sdk/build.gradle.kts [hunk 2]
```diff
@@ -15,32 +15,33 @@
...
-    dependsOn(it.tasks.named("copyApp"))
-    dependsOn(it.tasks.named("copyLib"))
-  }
-  dependsOn(project(":swirlds").tasks.named("copyApp"))
+    group = "application"
+    val sdkDir = File(rootProject.projectDir, "sdk")
+    workingDir = sdkDir
+    jvmArgs =
+        listOf(
+            "-agentlib:jdwp=transport=dt_socket,address=8888,server=y,suspend=n",
```
Comment (jsync-swirlds) [inline platform-sdk/build.gradle.kts:31]: May want to look at the PR #7264 (in progress) to remove Helidon.  That changes some of the requirements for the command line used here (it also changes the dependencies quite a lot).

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts [hunk 3]
```diff
@@ -15,31 +15,39 @@
  */
 
 plugins {
-  id("com.swirlds.platform.conventions")
-  id("com.swirlds.platform.application")
-  id("com.swirlds.platform.protobuf")
+    id("com.swirlds.platform.conventions")
+    id("com.swirlds.platform.application")
+    id("com.google.protobuf")
```
Comment (jsync-swirlds) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:20]: We should make sure this is correct.  We replaced google's protobuf compiler in the main codebase, so I would think the testing tool should use the same plugin.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts [hunk 4]
```diff
@@ -15,31 +15,39 @@
...
+    implementation(project(":swirlds-unit-tests:structures:swirlds-merkle-test"))
 
-  testImplementation(testLibs.bundles.junit)
-  testImplementation(testLibs.bundles.mocking)
+    testImplementation(testLibs.bundles.junit)
+    testImplementation(testLibs.bundles.mocking)
 }
 
-tasks.withType<Javadoc>() { enabled = false }
+protobuf { protoc { artifact = "com.google.protobuf:protoc:3.21.5" } }
```
Comment (jsync-swirlds) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:45]: We should make sure this is correct. We replaced google's protobuf compiler in the main codebase, so I would think the testing tool should use the same plugin.

**Review by hendrikebbers — DISMISSED** [review]

**Review by timo0 — APPROVED** [review]

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts (hunk 3 above)
Comment (jjohannes) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:20]: Right now, this is preserving what is on `develop`. Just moved things around.

Maybe we can address this concern in a follow up once this is merged?

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts (hunk 4 above)
Comment (jjohannes) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:45]: Right now, this is preserving what is on `develop`. Just moved things around.

Maybe we can address this concern in a follow up once this is merged?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java [hunk 5]
```diff
@@ -6,19 +6,18 @@
     requires transitive com.hedera.node.config;
     requires transitive com.hedera.node.hapi;
     requires transitive com.hedera.pbj.runtime;
+    requires transitive com.swirlds.common;
+    requires transitive com.swirlds.config;
```
Comment (jsync-swirlds) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java:10]: What does this service expose in its API that requires config?
With the amount of transitives we're adding, it is becoming important to be sure we actually need transitive for all of them.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts (hunk 3 above)
Comment (jsync-swirlds) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:20]: It is currently `com.swirlds.platform.protobuf`, and this PR changes to `com.google.protobuf`.  That's different from develop, so I'm asking if that's correct, given we don't directly use google's protobuf implementation with the switch to PBJ.

Comment (nathanklick) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:20]: @jsync-swirlds PTT has never used PBJ and has always used the official google protobuf implementation and compiler. @jjohannes changes here are correct. The removal of the gradle `id("com.swirlds.platform.protobuf")` custom plugin is due to consolidating the changes into the `build.gradle.kts` definition of the project.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts (hunk 4 above)
Comment (nathanklick) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:45]: @jsync-swirlds I doubt we will ever convert PTT to use PBJ. PTT is in the process of being deprecated and removed. Additionally, PTT uses a small handful of PTT specific protobuf definitions.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java (hunk 5 above)
Comment (jjohannes) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java:10]: I did the new dependency analysis – `checkAllModuleInfo` – on this branch and addressed all things reported. I did that to have a "clean" state to check that everything is fine with the changes in the module patching we did.

```
./gradlew app-service-contract-impl:reason --id com.swirlds:swirlds-config 


----------------------------------------
You asked about the dependency 'com.swirlds:swirlds-config-api:0.40.0-adhoc.x5482f0a8'.
There is no advice regarding this dependency.
----------------------------------------

...

Source: main
------------
* Exposes 1 class: com.swirlds.config.api.Configuration (implies api).
```

I believe this is because:
```
com.hedera.node.app.service.contract.impl.hevm
```

is an exported package and is uses `com.swirlds.config.api.Configuration` in a [public signature here](https://github.com/hashgraph/hedera-services/blob/309d81511549804da5e7d4269942695ea5e4c66a/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionProcessor.java#L39).

-----

We don't know yet how we want to work with this check on the long run. If you prefer, I can revert this change as it is not needed for this branch. It should probably be addressed in the code by someone knowing this area instead of me (who has no idea about the code itself) changing dependencies.

Comment (jjohannes) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java:9]: To illustrate this, I addressed what made `com.swirlds.common` transitive, as I am quite confident with that change:
https://github.com/hashgraph/hedera-services/commit/0c6f6d65e259eb83ab2ca028e78d5ce7afb7fa89

Comment (jsync-swirlds) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/module-info.java:9]: I'm fine with the team reviewing and making changes to reduce the transitive requirements in a future PR.  Knowing that we have API leaks like this is sufficient clarification for now.

**Review by nathanklick — DISMISSED** [review]

**Comment by nathanklick** [comment]
Conflicts will need to be resolved prior to merging; otherwise, it looks good to me.

**Review by jsync-swirlds — DISMISSED** [review]

**Review by nathanklick — APPROVED** [review]

File: build-logic/src/main/kotlin/com.hedera.hashgraph.jpms-module-dependencies.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR
+ * DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES.
+ */
+
+plugins {
+    id("org.gradlex.java-module-dependencies")
+}
+
+javaModuleDependencies {
+    warnForMissingVersions.set(false) // do not expect versions in catalog
```
Comment (hendrikebbers) [inline build-logic/src/main/kotlin/com.hedera.hashgraph.jpms-module-dependencies.gradle.kts:23]: What does that flag do?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmBlocks.java [hunk 7]
```diff
@@ -48,7 +48,7 @@ public interface HederaEvmBlocks {
      */
     BlockValues blockValuesOf(long gasLimit);
 
-    static Hash besuHashFrom(@NonNull final com.swirlds.common.crypto.Hash hash) {
+    private static Hash besuHashFrom(@NonNull final com.swirlds.common.crypto.Hash hash) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmBlocks.java:51]: Is that change needed for the PR?

**Review by hendrikebbers — APPROVED** [review]

File: platform-sdk/build.gradle.kts [hunk 8]
```diff
@@ -15,32 +15,33 @@
  */
 
 plugins {
-  id("com.swirlds.platform.aggregate-reports")
-  id("com.swirlds.platform.spotless-conventions")
-  id("com.swirlds.platform.spotless-kotlin-conventions")
+    id("com.swirlds.platform.aggregate-reports")
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply platform-sdk/build.gradle.kts:18]: This will be addressed in the follow up. Started here: #7875

My proposal is:
- For now: get rid of the `com.swirlds.platform` prefix here and use `com.hedera.hashgraph` for all the convention plugins (will complete this transition in follow up PR)
- Later: Rename `com.hedera.hashgraph` to whatever you want that to be in the future

File: platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts [hunk 9]
```diff
@@ -44,4 +44,12 @@ dependencies {
     testImplementation(testFixtures(project(":swirlds-common")))
 }
 
-tasks.withType<Javadoc>() { enabled = false }
+protobuf { protoc { artifact = "com.google.protobuf:protoc:3.21.5" } }
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts:47]: This is something specific to the "protobuf" plugin. But I can check this again. Will do in the follow up #7875, where I would like to do more streamlining of the "platform-apps" demo/example applications setup anyway.

File: settings.gradle.kts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjohannes) [reply settings.gradle.kts:39]: See: https://github.com/hashgraph/hedera-services/pull/7330#discussion_r1246756195

This is the switch to turn on the "true" integration between "hedera-node" and "platform-sdk". Once this is on,  "hedera-node" will build against the sources of "platform-sdk" (instead of the published version). 

I commented it out in this PR, because the code (right now) won't compile. There are changes in "platform-sdk" right now that are not compatible.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmBlocks.java (hunk 7 above)
Comment (jjohannes) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmBlocks.java:51]: Not necessarily, but I wanted the dependency check to pass. And this was also an example of what you can detect with the check. See these comments: https://github.com/hashgraph/hedera-services/pull/7330#discussion_r1258458493



Note: strictly speaking the dependency check should not have complained here. It should **not** consider package-private things to be public API in a Module context. Because there, a package is always only accessible within the module. This is different to classpath-based Java. I already took a note to provide a fix for this in the analysis plugin.

File: build-logic/src/main/kotlin/com.hedera.hashgraph.jpms-module-dependencies.gradle.kts (hunk 6 above)
Comment (jjohannes) [reply build-logic/src/main/kotlin/com.hedera.hashgraph.jpms-module-dependencies.gradle.kts:23]: As soon as a version catalog is available, the module-dependencies plugin automatically looks for versions in it. If it can't find a version, it warns. This turns this behavior off.

We need it, because we still use the catalog right now for a few things unrelated to modules. If we won't use the catalog (in settings.gradle.kts) at all, the warnings would be turned off by default.

I should add more descriptive comments to all of these things in a follow up PR.

**Review by netopyr — APPROVED** [review]
LGTM
